# News Headlines Sentiment

Use the news api to pull the latest news articles for bitcoin and ethereum and create a DataFrame of sentiment scores for each coin. 

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [687]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
load_dotenv()
from newsapi import NewsApiClient
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer
nltk.download('vader_lexicon')
analyzer = SentimentIntensityAnalyzer()


%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [688]:
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/draganbogatic/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [689]:
# Read your api key environment variable
api_key = os.getenv("NEWS_API_KEY")

In [690]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=api_key)

In [691]:
# Fetch the Bitcoin news articles
btc_headlines = newsapi.get_everything(q="bitcoin", language="en", page_size=100,sort_by="relevancy")
btc_articles = btc_headlines["articles"]
btc_articles[0]

{'source': {'id': 'wired', 'name': 'Wired'},
 'author': 'Timothy B. Lee, Ars Technica',
 'title': 'An Engineer Gets 9 Years for Stealing $10M From Microsoft',
 'description': 'The defendant tried—and failed—to use bitcoin to cover his tracks.',
 'url': 'https://www.wired.com/story/an-engineer-gets-9-years-for-stealing-dollar10m-from-microsoft/',
 'urlToImage': 'https://media.wired.com/photos/5fac6afb446b4639b3d5b8d8/191:100/w_1280,c_limit/Security-Microsoft-1229426260.jpg',
 'publishedAt': '2020-11-12T14:00:00Z',
 'content': "A former Microsoft software engineer from Ukraine has been sentenced to nine years in prison for stealing more than $10 million in store credit from Microsoft's online store. From 2016 to 2018, Volod… [+3307 chars]"}

In [692]:
# Fetch the Ethereum news articles
eth_headlines = newsapi.get_everything(q="ethereum", language="en", page_size=100,sort_by="relevancy")
eth_articles = eth_headlines["articles"]
eth_articles[0]

{'source': {'id': 'engadget', 'name': 'Engadget'},
 'author': 'Karissa Bell',
 'title': 'PayPal now lets all US users buy, sell and hold cryptocurrency',
 'description': 'PayPal is bringing its newly-announced support for cryptocurrency to all US accounts. It first announced plans to open cryptocurrency trading to US-based users in October, but until now it was only available to a small subset of PayPal account holders. That’s…',
 'url': 'https://www.engadget.com/paypal-opens-cryptocurrency-all-us-accounts-210541778.html',
 'urlToImage': 'https://o.aolcdn.com/images/dims?resize=1200%2C630&crop=1200%2C630%2C0%2C0&quality=95&image_uri=https%3A%2F%2Fs.yimg.com%2Fos%2Fcreatr-uploaded-images%2F2020-11%2F4e1af080-2528-11eb-bbdf-193e571d142a&client=amp-blogside-v2&signature=22ad23dde6ee3e667a8067a4c39e1962659cec92',
 'publishedAt': '2020-11-12T21:05:41Z',
 'content': 'PayPal is bringing its newly-announced support for cryptocurrency to all US accounts. It first announced plans to open cryptoc

In [693]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiments = []

for article in btc_articles:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        btc_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
# Create DataFrame
btc_df = pd.DataFrame(btc_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
btc_df = btc_df[cols]

In [694]:
# Drop duplicates articles
btc_df.drop_duplicates(keep="first", inplace=True)
btc_df.head()

date                                               text  compound  \
0  2020-11-12  A former Microsoft software engineer from Ukra...   -0.6705   
1  2020-12-03  Visa has partnered with cryptocurrency startup...    0.6369   
2  2020-11-12  PayPal is bringing its newly-announced support...    0.2144   
3  2020-11-20  In November 2017, after an absolutely massive,...    0.2023   
4  2020-12-06  Unlike ‘conventional’ cryptocurrencies, a cent...    0.0000   

   positive  negative  neutral  
0     0.064     0.199    0.737  
1     0.162     0.000    0.838  
2     0.053     0.000    0.947  
3     0.050     0.000    0.950  
4     0.000     0.000    1.000

In [695]:
# Create the ethereum sentiment scores DataFrame
eth_sentiments = []

for article in eth_articles:
    try:
        text = article["content"]
        date = article["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment["compound"]
        pos = sentiment["pos"]
        neu = sentiment["neu"]
        neg = sentiment["neg"]
        
        eth_sentiments.append({
            "text": text,
            "date": date,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu
            
        })
        
    except AttributeError:
        pass
    
# Create DataFrame
eth_df = pd.DataFrame(eth_sentiments)

# Reorder DataFrame columns
cols = ["date", "text", "compound", "positive", "negative", "neutral"]
eth_df = eth_df[cols]

In [696]:
# Drop duplicates articles
eth_df.drop_duplicates(keep="first", inplace=True)
eth_df.head()

date                                               text  compound  \
0  2020-11-12  PayPal is bringing its newly-announced support...    0.2144   
1  2020-11-23  FILE PHOTO: Representation of the Ethereum vir...    0.0000   
2  2020-11-23  LONDON (Reuters) - Digital currencies Ethereum...    0.4215   
4  2020-11-19  PayPal has launched the Generosity Network, a ...    0.8779   
5  2020-11-26  By Reuters Staff\r\nFILE PHOTO: A representati...    0.0000   

   positive  negative  neutral  
0     0.053       0.0    0.947  
1     0.000       0.0    1.000  
2     0.088       0.0    0.912  
4     0.318       0.0    0.682  
5     0.000       0.0    1.000

In [697]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound   positive   negative    neutral
count  87.000000  87.000000  87.000000  87.000000
mean    0.165337   0.061207   0.020655   0.918126
std     0.343146   0.066855   0.043497   0.077008
min    -0.670500   0.000000   0.000000   0.682000
25%     0.000000   0.000000   0.000000   0.859000
50%     0.094600   0.050000   0.000000   0.935000
75%     0.458550   0.121500   0.017500   1.000000
max     0.877900   0.318000   0.215000   1.000000

In [698]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound   positive   negative    neutral
count  95.000000  95.000000  95.000000  95.000000
mean    0.222034   0.074853   0.022695   0.902463
std     0.368386   0.077264   0.045878   0.086894
min    -0.670500   0.000000   0.000000   0.653000
25%     0.000000   0.000000   0.000000   0.849500
50%     0.226300   0.075000   0.000000   0.912000
75%     0.502250   0.126000   0.013500   1.000000
max     0.883400   0.347000   0.196000   1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: ETH has a mean positive score of 0.077646 compared to BTC mean positive score of 0.057571

Q: Which coin had the highest compound score?

A: Both BTC and ETH have the same Max compound score of 0.946800

Q. Which coin had the highest positive score?

A: ETH has the highest positive score of 0.347000 compared t0 BTC 0.297000

---

# Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word
2. Remove Punctuation
3. Remove Stopwords

In [699]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import regex
import re

In [700]:
btc_text_df = btc_df['text'].astype('string')
btc_text_df

0     A former Microsoft software engineer from Ukra...
1     Visa has partnered with cryptocurrency startup...
2     PayPal is bringing its newly-announced support...
3     In November 2017, after an absolutely massive,...
4     Unlike ‘conventional’ cryptocurrencies, a cent...
                            ...                        
93    For anyone serious about mining cryptocurrency...
94    PayPal has launched the Generosity Network, a ...
95    A woman walks past a Microsoft Store on 5th av...
96    November
18, 2020
2 min read
This article w...
97    FILE PHOTO: A sign of China Construction Bank ...
Name: text, Length: 87, dtype: string

In [701]:
eth_text_df = eth_df['text'].astype('string')
eth_text_df

0     PayPal is bringing its newly-announced support...
1     FILE PHOTO: Representation of the Ethereum vir...
2     LONDON (Reuters) - Digital currencies Ethereum...
4     PayPal has launched the Generosity Network, a ...
5     By Reuters Staff
FILE PHOTO: A representation...
                            ...                        
92    What makes an asset "liquid"? Liquidity basica...
93    Hybrid blockchain platform Kadena is teaming u...
94    Ethereum
Ethereum surged by 9.43% on Friday. ...
95    CoinDesk just snagged a front-row seat to a se...
96    Ethereum co-founder Vitalik Buterin, dressed l...
Name: text, Length: 95, dtype: string

In [702]:
# Complete the tokenizer function
def tokenizer(text):

    """Tokenizes text."""

    # Create a list of the words
    sentences = [sent_tokenize(article) for article in text]
    words = [word_tokenize(article) for article in text]

    # Remove punctuation using regex
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)
    words_clean = word_tokenize(re_clean)
   
   # Remove the stop words
    sw = set(stopwords.words('english'))
    #first_pass = [word for word in words_clean if word not in sw]

    # Define additional stopwords
    sw_addon = {'char', 'chars' 'la', 'im','volod', 'drin', 'wa', 'ha'}
    second_pass = [word.lower() for word in words_clean if word.lower() not in sw.union(sw_addon)] 

    # Lemmatize Words into root words; convert the words to lowercase
    lemmatizer = WordNetLemmatizer()
    lem = [lemmatizer.lemmatize(word) for word in words_clean]
    tokens = [word.lower() for word in lem if word.lower() not in sw.union(sw_addon)] 
  
    return tokens

In [703]:
# Tokenizing article in the forst row of the "text" column
tokenizer(text_df[0])

['former',
 'microsoft',
 'software',
 'engineer',
 'ukraine',
 'sentenced',
 'nine',
 'year',
 'prison',
 'stealing',
 'million',
 'store',
 'credit',
 'microsofts',
 'online',
 'store']

In [704]:
# Create for loop to iterate through the rows in the "text" column 
btc_tokens = []
for row in btc_text_df:
    token = tokenizer(row)
    btc_tokens.append(token)

    print(btc_tokens) 

'poised', 'end', 'week', 'loss', 'euro', 'yen', 'little', 'changed', 'v', 'dollar', 'bitcoin', 'hit', 'fresh', 'threeyear', 'high', 'graphic', 'world', 'fx', 'ra'], ['posted', 'bitcoin', 'soared', 'record', 'high', 'dollar', 'monday', 'rally', 'steamed', 'ahead', 'boosted', 'increased', 'demand', 'institutional', 'retail', 'investor', 'saw', 'virtual', 'c'], ['reuters', 'staffreuters', 'sticky', 'talksfile', 'photo', 'small', 'toy', 'figure', 'seen', 'front', 'brexit', 'logo', 'illustration', 'picture', 'march', 'reutersdado', 'ruvicillustrationbritai'], ['currency', 'asset', 'perhaps', 'commodity', 'effort', 'drive', 'price', 'infamous', 'crypto', 'evangelicals', 'coming', 'attributing', 'increasing', 'number', 'use', 'case'], ['new', 'zealand', 'dollar', 'set', 'reach', 'twoyear', 'high', 'commoditylinked', 'currency', 'gain', 'v', 'us', 'dollar', 'canadian', 'dollar', 'shy', 'twoyear', 'high', 'graphic', 'world', 'fx', 'rate', 'tmsnrtrsr'], ['fixes', 'typo', 'euro', 'price', 'th', '

In [705]:
# Create a new tokens column for bitcoin
btc_df['btc_token'] = btc_tokens
btc_df

date                                               text  compound  \
0   2020-11-12  A former Microsoft software engineer from Ukra...   -0.6705   
1   2020-12-03  Visa has partnered with cryptocurrency startup...    0.6369   
2   2020-11-12  PayPal is bringing its newly-announced support...    0.2144   
3   2020-11-20  In November 2017, after an absolutely massive,...    0.2023   
4   2020-12-06  Unlike ‘conventional’ cryptocurrencies, a cent...    0.0000   
..         ...                                                ...       ...   
93  2020-11-18  For anyone serious about mining cryptocurrency...    0.2023   
94  2020-11-19  PayPal has launched the Generosity Network, a ...    0.8779   
95  2020-11-13  A woman walks past a Microsoft Store on 5th av...   -0.5267   
96  2020-11-18  November\r\n18, 2020\r\n2 min read\r\nThis art...   -0.3400   
97  2020-11-23  FILE PHOTO: A sign of China Construction Bank ...    0.3182   

    positive  negative  neutral  \
0      0.064     0.199    0.737   
1      0.162     0.000    0.838   
2      0.053     0.000    0.947   
3      0.050     0.000    0.950   
4      0.000     0.000    1.000   
..       ...       ...      ...   
93     0.056     0.035    0.909   
94     0.318     0.000    0.682   
95     0.000     0.136    0.864   
96     0.000     0.064    0.936   
97     0.065     0.000    0.935   

                                            btc_token  
0   [former, microsoft, software, engineer, ukrain...  
1   [visahas, partnered, cryptocurrency, startup, ...  
2   [paypal, bringing, newlyannounced, support, cr...  
3   [november, absolutely, massive, twomonth, rall...  
4   [unlike, conventional, cryptocurrencies, centr...  
..                                                ...  
93  [anyone, serious, mining, cryptocurrency, bitc...  
94  [paypal, launched, generosity, network, platfo...  
95  [woman, walk, past, microsoft, store, th, aven...  
96  [november, min, readthis, article, translated,...  
97  [file, photo, sign, china, construction, bank,...  

[87 rows x 7 columns]

In [706]:
eth_tokens = []
for row in eth_text_df:
    token = tokenizer(row)
    eth_tokens.append(token)

    print(eth_tokens) 

e', 'secondlargest', 'donation'], ['make', 'asset', 'liquid', 'liquidity', 'basically', 'refers', 'item', 'marketability', 'faster', 'convert', 'something', 'cash', 'value', 'retains', 'sell', 'liquid'], ['hybrid', 'blockchain', 'platform', 'kadena', 'teaming', 'stablecoin', 'maker', 'terra', 'aim', 'expanding', 'decentralized', 'finance', 'defi', 'platformannounced', 'tuesday', 'kadena', 'said', 'add'], ['ethereumethereum', 'surged', 'friday', 'following', 'rally', 'thursday', 'ethereum', 'ended', 'day', 'particularly', 'bullish', 'day', 'ethereum', 'rallied', 'early', 'mornin'], ['coindesk', 'snagged', 'frontrow', 'seat', 'seminal', 'event', 'crypto', 'industry', 'ticket', 'price', 'ethas', 'medium', 'outlet', 'covering', 'nascent', 'technology', 'believe', 'sometimes', 'chronicle']]
[['paypal', 'bringing', 'newlyannounced', 'support', 'cryptocurrency', 'us', 'account', 'first', 'announced', 'plan', 'open', 'cryptocurrency', 'trading', 'usbased', 'user', 'october'], ['file', 'photo',

In [707]:
# Create a new tokens column for ethereum
eth_df['eth_token'] = eth_tokens
eth_df

date                                               text  compound  \
0   2020-11-12  PayPal is bringing its newly-announced support...    0.2144   
1   2020-11-23  FILE PHOTO: Representation of the Ethereum vir...    0.0000   
2   2020-11-23  LONDON (Reuters) - Digital currencies Ethereum...    0.4215   
4   2020-11-19  PayPal has launched the Generosity Network, a ...    0.8779   
5   2020-11-26  By Reuters Staff\r\nFILE PHOTO: A representati...    0.0000   
..         ...                                                ...       ...   
92  2020-11-12  What makes an asset "liquid"? Liquidity basica...    0.6361   
93  2020-11-10  Hybrid blockchain platform Kadena is teaming u...    0.0000   
94  2020-11-07  Ethereum\r\nEthereum surged by 9.43% on Friday...    0.0000   
95  2020-11-24  CoinDesk just snagged a front-row seat to a se...    0.0000   
96  2020-11-22  Ethereum co-founder Vitalik Buterin, dressed l...    0.3612   

    positive  negative  neutral  \
0      0.053       0.0    0.947   
1      0.000       0.0    1.000   
2      0.088       0.0    0.912   
4      0.318       0.0    0.682   
5      0.000       0.0    1.000   
..       ...       ...      ...   
92     0.132       0.0    0.868   
93     0.000       0.0    1.000   
94     0.000       0.0    1.000   
95     0.000       0.0    1.000   
96     0.075       0.0    0.925   

                                            eth_token  
0   [paypal, bringing, newlyannounced, support, cr...  
1   [file, photo, representation, ethereum, virtua...  
2   [london, reuters, digital, currency, ethereum,...  
4   [paypal, launched, generosity, network, platfo...  
5   [reuters, stafffile, photo, representation, vi...  
..                                                ...  
92  [make, asset, liquid, liquidity, basically, re...  
93  [hybrid, blockchain, platform, kadena, teaming...  
94  [ethereumethereum, surged, friday, following, ...  
95  [coindesk, snagged, frontrow, seat, seminal, e...  
96  [ethereum, cofounder, vitalik, buterin, dresse...  

[95 rows x 7 columns]

---

# NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [708]:
from collections import Counter
from nltk import ngrams

In [744]:
# Create one list of tokens from list of tokens in each row
btc_processed = []
for row in btc_df['btc_token']:
    for item in row:
        btc_processed.append(item)
        print(btc_processed)

lar', 'poised', 'end', 'week', 'loss', 'euro', 'yen', 'little', 'changed', 'v', 'dollar', 'bitcoin', 'hit', 'fresh', 'threeyear', 'high', 'graphic', 'world', 'fx', 'ra', 'posted', 'bitcoin', 'soared', 'record', 'high', 'dollar', 'monday', 'rally', 'steamed', 'ahead', 'boosted', 'increased', 'demand', 'institutional', 'retail', 'investor', 'saw', 'virtual', 'c', 'reuters', 'staffreuters', 'sticky', 'talksfile', 'photo', 'small', 'toy', 'figure', 'seen', 'front', 'brexit', 'logo', 'illustration', 'picture', 'march', 'reutersdado', 'ruvicillustrationbritai', 'currency', 'asset', 'perhaps', 'commodity', 'effort', 'drive', 'price', 'infamous', 'crypto', 'evangelicals', 'coming', 'attributing', 'increasing', 'number', 'use', 'case', 'new', 'zealand', 'dollar', 'set', 'reach', 'twoyear', 'high', 'commoditylinked', 'currency', 'gain', 'v', 'us', 'dollar', 'canadian', 'dollar', 'shy', 'twoyear', 'high', 'graphic', 'world', 'fx', 'rate', 'tmsnrtrsr', 'fixes', 'typo', 'euro', 'price', 'th', 'para

In [745]:
eth_processed = []
for row in eth_df['eth_token']:
    for item in row:
        eth_processed.append(item)
        print(eth_processed)

'ftx', 'made', 'secondlargest', 'donation', 'make', 'asset', 'liquid', 'liquidity', 'basically', 'refers', 'item', 'marketability', 'faster', 'convert', 'something', 'cash', 'value', 'retains', 'sell', 'liquid', 'hybrid', 'blockchain', 'platform', 'kadena', 'teaming', 'stablecoin', 'maker', 'terra', 'aim', 'expanding', 'decentralized', 'finance', 'defi', 'platformannounced', 'tuesday', 'kadena', 'said', 'add', 'ethereumethereum', 'surged', 'friday', 'following', 'rally', 'thursday', 'ethereum', 'ended', 'day', 'particularly', 'bullish', 'day', 'ethereum', 'rallied', 'early', 'mornin', 'coindesk', 'snagged', 'frontrow', 'seat', 'seminal', 'event', 'crypto', 'industry', 'ticket', 'price', 'ethas', 'medium', 'outlet', 'covering', 'nascent', 'technology', 'believe', 'sometimes', 'chronicle', 'ethereum', 'cofounder', 'vitalik', 'buterin', 'dressed', 'like', 'medieval', 'harlequin', 'leaning', 'velvet', 'chair', 'stare', 'past', 'viewer', 'distance', 'rocketed', 'sky', 'ether']
['paypal', 'b

In [739]:
# Generate the Bitcoin N-grams where N=2
bigram_counts = Counter(ngrams(btc_processed, n=2))
btc_bigram.append(bigram_counts)

print(dict(bigram_counts))

'niche'): 3, ('niche', 'asset'): 3, ('asset', 'take'): 3, ('take', 'punt'): 3, ('punt', 'bitcoin'): 3, ('bitcoin', 'headline'): 3, ('headline', 'rekindling'): 3, ('rekindling', 'debate'): 3, ('debate', 'true'): 3, ('true', 'role'): 3, ('role', 'driver'): 3, ('driver', 'behind'): 3, ('behind', 'london'): 1, ('behind', 'november'): 1, ('november', 'min'): 3, ('min', 'readopinions'): 1, ('readopinions', 'expressed'): 1, ('expressed', 'entrepreneur'): 1, ('entrepreneur', 'contributor'): 1, ('contributor', 'ownsquare'): 1, ('ownsquare', 'inc'): 1, ('inc', 'payment'): 1, ('payment', 'platform'): 1, ('platform', 'headed'): 1, ('headed', 'famous'): 1, ('famous', 'twitter'): 1, ('twitter', 'ceo'): 1, ('ceo', 'jack'): 1, ('jack', 'dorseyhas'): 1, ('dorseyhas', 'announced'): 1, ('announced', 'october'): 1, ('october', 'reuters'): 1, ('reutersdado', 'ruvicillustrationnew'): 2, ('ruvicillustrationnew', 'york'): 2, ('bitcoin', 'btcbtsp'): 2, ('btcbtsp', 'rose'): 2, ('rose', 'n'): 2, ('n', 'reuters')

In [746]:
# Generate the Ethereum N-grams where N=2
bigram_counts = Counter(ngrams(eth_processed, n=2))
print(dict(bigram_counts)) 

 'transaction'): 2, ('transaction', 'growth'): 1, ('growth', 'positive'): 1, ('positive', 'indicator'): 1, ('indicator', 'future'): 1, ('future', 'financial'): 1, ('financial', 'use'): 1, ('use', 'casesullibitcoin'): 1, ('casesullibitcoin', 'btc'): 1, ('btc', 'trading'): 2, ('trading', 'around'): 2, ('around', 'consensys'): 1, ('consensys', 'adding'): 1, ('adding', 'dapp'): 1, ('dapp', 'development'): 1, ('development', 'platform'): 1, ('platform', 'truffle'): 1, ('truffle', 'suite'): 1, ('suite', 'back'): 1, ('back', 'lineup'): 1, ('lineup', 'ethereum'): 1, ('ethereum', 'toolsconsensys'): 1, ('toolsconsensys', 'chief'): 1, ('chief', 'strategy'): 1, ('strategy', 'officer'): 1, ('officer', 'simon'): 1, ('simon', 'morris'): 1, ('morris', 'said'): 1, ('said', 'tuesday'): 1, ('tuesday', 'ethereumfocused'): 1, ('ethereumfocused', 'firm'): 1, ('firm', 'acquired'): 1, ('acquired', 'th'): 1, ('th', 'idex'): 1, ('idex', 'betting'): 1, ('betting', 'three'): 1, ('three', 'chain'): 1, ('chain', 'b

In [711]:
# Use the token_count function to generate the top 10 words from each coin
def token_count(tokens, N=10):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [748]:
# Get the top 10 words for Bitcoin
btc_top_ten = token_count(btc_processed)
btc_top_ten

[('bitcoin', 69),
 ('reuters', 55),
 ('currency', 32),
 ('virtual', 24),
 ('taken', 23),
 ('photo', 20),
 ('illustration', 19),
 ('reutersdado', 19),
 ('cryptocurrency', 18),
 ('seen', 18)]

In [749]:
# Get the top 10 words for Ethereum
eth_top_ten = token_count(eth_processed)
eth_top_ten

[('ethereum', 55),
 ('bitcoin', 39),
 ('reuters', 17),
 ('blockchain', 17),
 ('cryptocurrency', 14),
 ('price', 13),
 ('first', 12),
 ('currency', 11),
 ('asset', 11),
 ('november', 10)]

# Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [715]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [716]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [717]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

# Named Entity Recognition

In this section, you will build a named entity recognition model for both coins and visualize the tags using SpaCy.

In [718]:
import spacy
from spacy import displacy

In [719]:
# Optional - download a language model for SpaCy
# !python -m spacy download en_core_web_sm

In [720]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

## Bitcoin NER

In [721]:
# Concatenate all of the bitcoin text together
# YOUR CODE HERE!

In [722]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [723]:
# Render the visualization
# YOUR CODE HERE!

In [724]:
# List all Entities
# YOUR CODE HERE!

---

## Ethereum NER

In [725]:
# Concatenate all of the bitcoin text together
# YOUR CODE HERE!

In [726]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [727]:
# Render the visualization
# YOUR CODE HERE!

In [728]:
# List all Entities
# YOUR CODE HERE!